src/operator/nn/mkldnn/mkldnn_base.cc

```diff
@@ -77,26 +77,35 @@ mkldnn::memory *TmpMemMgr::Alloc(const mkldnn::memory::primitive_desc &pd) {
   }
 }
 
-mkldnn_output_t CreateMKLDNNMem(const NDArray &arr,
-                                const mkldnn::memory::primitive_desc &desc,
-                                OpReqType req) {
+bool CanWriteTo(const NDArray &out_arr,
+                const NDArray* in_arr,
+                const mkldnn::memory::primitive_desc &desc) {
+  bool add_same = in_arr->GetMKLDNNData()->get_data_handle() ==
+      out_arr.GetMKLDNNData()->get_data_handle();
+  bool pdesc_same = out_arr.GetMKLDNNData()->get_primitive_desc() == desc;
+  return add_same && pdesc_same;
+}
+
+mkldnn_output_t CreateMKLDNNMem(const NDArray &out_arr,
+                                 const mkldnn::memory::primitive_desc &desc,
+                                 OpReqType req, const NDArray* in_arr) {
   if (kAddTo == req) {
     auto tmp = TmpMemMgr::Get()->Alloc(desc);
     return mkldnn_output_t(OutDataOp::AddBack, tmp);
-  } else if (kWriteInplace == req) {
-    // MKLDNN ops may not support the case that the input and the output uses
-    // the same memory. Let's use an extra copy to make sure it always works.
-    auto tmp = TmpMemMgr::Get()->Alloc(desc);
-    return mkldnn_output_t(OutDataOp::CopyBack, tmp);
-  } else {
-    mkldnn::memory *mem = const_cast<NDArray &>(arr).CreateMKLDNNData(desc);
-    if (mem == nullptr) {
-      auto tmp = TmpMemMgr::Get()->Alloc(desc);
-      return mkldnn_output_t(OutDataOp::CopyBack, tmp);
-    } else {
+  } else if (req == kWriteInplace) {
+    if (in_arr != nullptr && CanWriteTo(out_arr, in_arr, desc)) {
+      mkldnn::memory *mem = const_cast<NDArray &>(out_arr).CreateMKLDNNData(desc);
       return mkldnn_output_t(OutDataOp::Noop, mem);
     }
+    auto tmp = TmpMemMgr::Get()->Alloc(desc);
+    return mkldnn_output_t(OutDataOp::CopyBack, tmp);
+  }
+  mkldnn::memory *mem = const_cast<NDArray &>(out_arr).CreateMKLDNNData(desc);
+  if (mem == nullptr) {
```

Comment:
so if `req == kNullOp`, code will still run to here? BTW, I would preper `nullptr == mem`.

Reply:
if req == kNullOp shouldn't we break before trying to create memory?

Reply:
I see. I'm just wondering if we need put all illegal req values into an else branch. If we set a unit test for this function, i guess any req value can be passed into this function, including kNullOp, without correctly handled.